### ISSUE TITLE
Bug: possible regression of TypeOrValueSpecifier for scoped packages with restrict-template-expressions

### ISSUE THREAD

**Issue by simon-v-swyftx** [typescript-eslint/typescript-eslint#10038]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

I was using the new `allow` option for `restrict-template-expressions` that [landed in `8.6.0`](https://github.com/typescript-eslint/typescript-eslint/pull/8556)

I expected that:

```typescript
// packages/flat-config-typed-tsconfig/src/index.ts
import { Uuid } from '@company/core'

export class Service {
  public createPosition(
    id: Uuid,
  ): void {
    // Invalid type "Uuid" of template literal expression
    const a = `someUrl/${id}`;
                      //  ^ specifically not working when specifying package: @company/core
  }
}
```
does **not** report [@typescript-eslint/restrict-template-expressions](https://typescript-eslint.io/rules/restrict-template-expressions) with the following config:

```javascript
"@typescript-eslint/restrict-template-expressions": [
  "error",
  {
    allow: [
      // doesn't appear to be working
      {
        from: "package",
        name: "Uuid",
        package: "@company/core",
      },
    ],
  },
],
```

The uuid file in question exports a class with a custom `.toString` so I don't want it to error 🙂 See reproduction repo

Feels related to this old fix: https://github.com/typescript-eslint/typescript-eslint/pull/6780 but I might be barking up the wrong tree

### Reproduction Repository Link

https://github.com/simon-v-swyftx/typescript-eslint-reproducable

### Repro Steps

1) clone the repo
2) `yarn install`
3) view `packages/flat-config-typed-tsconfig/src/index.ts`
4) observe error [on line 8](https://github.com/simon-v-swyftx/typescript-eslint-reproducable/blob/main/packages/flat-config-typed-tsconfig/src/index.ts#L8)

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.6.0` |
| `@typescript-eslint/parser`            | `8.6.0` |
| `@typescript-eslint/scope-manager`     | `8.6.0` |
| `@typescript-eslint/typescript-estree` | `8.6.0` |
| `@typescript-eslint/type-utils`        | `8.6.0` |
| `@typescript-eslint/utils`             | `8.6.0` |
| `TypeScript`                           | `5.6.2` |
| `ESLint`                               | `9.11.0` |
| `node`                                 | `20.10.0` |

**Comment by auvred** [comment]
> 2\. yarn install

It fails with the following error (I assumed the yarn should be `v3`. if so, it would be nice to specify `packageManager` in the `package.json`)

<details>

<summary>
yarn install logs
</summary>

```
> yarn
! Corepack is about to download https://registry.yarnpkg.com/yarn/-/yarn-1.22.22.tgz
? Do you want to continue? [Y/n]

! The local project doesn't define a 'packageManager' field. Corepack will now add one referencing yarn@1.22.22+sha1.ac34549e6aa8e7ead463a7407e1c7390f61a6610.
! For more details about this field, consult the documentation at https://nodejs.org/api/packages.html#packagemanager

yarn install v1.22.22
warning Missing version in workspace at "/tmp/repro/typescript-eslint-reproducable/packages/core", ignoring.
[1/4] Resolving packages...
error Couldn't find package "@company/core@workspace:*" required by "flat-config-typed-tsconfig@0.0.0" on the "npm" registry.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

</details>

---

The result of a little investigation: the issue is that typescript doesn't have `@company/core` in the `program.sourceFileToPackageName` 
https://github.com/typescript-eslint/typescript-eslint/blob/896c7312dfe8374312df9404cb11db8378cd091c/packages/type-utils/src/typeOrValueSpecifiers/typeDeclaredInPackageDeclarationFile.ts#L38j

Looks like TS has a special behavior for locally linked workspace packages.
